SimMuon/GEMDigitizer/src/GEMSignalModel.cc

```diff
@@ -0,0 +1,151 @@
+                              const edm::PSimHitContainer& simHits,
+                              CLHEP::HepRandomEngine* engine,
+                              Strips& strips_,
+                              DetectorHitMap& detectorHitMap_) {
+  bool digiMuon = false;
+  bool digiElec = false;
+  for (const auto& hit : simHits) {
+    if (std::abs(hit.particleType()) != 13 && digitizeOnlyMuons_)
+      continue;
+    double elecEff = 0.;
+    double partMom = hit.pabs();
+    double checkMuonEff = CLHEP::RandFlat::shoot(engine, 0., 1.);
+    double checkElecEff = CLHEP::RandFlat::shoot(engine, 0., 1.);
+    if (std::abs(hit.particleType()) == 13 && checkMuonEff < averageEfficiency_)
+      digiMuon = true;
+    if (std::abs(hit.particleType()) != 13)  //consider all non muon particles with gem efficiency to electrons
+    {
+      if (partMom <= 1.95e-03)
+        elecEff = 1.7e-05 * std::exp(2.1 * partMom * 1000.);
+      if (partMom > 1.95e-03 && partMom < 10.e-03)
+        elecEff =
+            1.34 * log(7.96e-01 * partMom * 1000. - 5.75e-01) / (1.34 + log(7.96e-01 * partMom * 1000. - 5.75e-01));
+      if (partMom > 10.e-03)
+        elecEff = 1.;
+      if (checkElecEff < elecEff)
+        digiElec = true;
+    }
+    if (!(digiMuon || digiElec))
+      continue;
+    const int bx(getSimHitBx(&hit, engine));
+    const std::vector<std::pair<int, int> >& cluster(simulateClustering(roll, &hit, bx, engine));
+    for (const auto& digi : cluster) {
+      detectorHitMap_.emplace(digi, &hit);
+      strips_.emplace(digi);
+    }
+  }
+}
+
+int GEMSignalModel::getSimHitBx(const PSimHit* simhit, CLHEP::HepRandomEngine* engine) {
+  int bx = -999;
+  const LocalPoint simHitPos(simhit->localPosition());
+  const float tof(simhit->timeOfFlight());
+  // random Gaussian time correction due to electronics jitter
+  float randomJitterTime = CLHEP::RandGaussQ::shoot(engine, 0., timeJitter_);
+  const GEMDetId id(simhit->detUnitId());
+  const GEMEtaPartition* roll(geometry_->etaPartition(id));
+  if (!roll) {
+    throw cms::Exception("Geometry") << "GEMSignalModel::getSimHitBx() - GEM simhit id does not match any GEM roll id: "
+                                     << id << "\n";
+    return 999;
+  }
+  if (roll->id().region() == 0) {
+    throw cms::Exception("Geometry")
+        << "GEMSignalModel::getSimHitBx() - this GEM id is from barrel, which cannot happen: " << roll->id() << "\n";
+  }
+  const double cspeed = 299792458;  // signal propagation speed in vacuum in [m/s]
+  const int nstrips = roll->nstrips();
+  float middleStrip = nstrips / 2.;
+  const LocalPoint& middleOfRoll = roll->centreOfStrip(middleStrip);
+  const GlobalPoint& globMiddleRol = roll->toGlobal(middleOfRoll);
+  double muRadius = sqrt(globMiddleRol.x() * globMiddleRol.x() + globMiddleRol.y() * globMiddleRol.y() +
+                         globMiddleRol.z() * globMiddleRol.z());
+  double timeCalibrationOffset_ = (muRadius * 1e+9) / (cspeed * 1e+2);  //[ns]
```

Comment:
more magic numbers

Reply:
@cvuosalo maybe this is a good place to use https://github.com/cms-sw/cmssw/blob/master/DataFormats/Math/interface/GeantUnits.h for unit conversions?

Reply:
Yes, if the CLHEP unit convention is in use, then `GeantUnits.h` matches that convention. CLHEP::c_light is in mm/ns. I assume `muRadius` is in cm. Then this line could be written:
```
// timeCalibrationOffset_ is in ns since c_light is mm/ns.
double timeCalibrationOffset_ = geant_units::operators::convertCmToMm(muRadius) / CLHEP::c_light;
```